core/src/main/java/org/elasticsearch/gateway/ReplicaShardAllocator.java

```diff
@@ -286,35 +342,57 @@ private MatchingNodes findMatchingNodes(ShardRouting shard, RoutingAllocation al
             // we only check for NO, since if this node is THROTTLING and it has enough "same data"
             // then we will try and assign it next time
             Decision decision = allocation.deciders().canAllocate(shard, node, allocation);
+
+            long matchingBytes = -1;
             if (explain) {
-                nodeDecisions.put(node.nodeId(), decision);
+                matchingBytes = computeMatchingBytes(primaryStore, storeFilesMetaData);
+                ShardStore shardStore;
+                if (matchingBytes == Long.MAX_VALUE) {
+                    shardStore = new ShardStore(StoreStatus.MATCHING_SYNC_ID, -1);
+                } else {
+                    shardStore = new ShardStore(StoreStatus.UNKNOWN, matchingBytes);
+                }
+                nodeDecisions.put(node.nodeId(), new NodeAllocationResult(discoNode, shardStore, decision));
             }
 
             if (decision.type() == Decision.Type.NO) {
                 continue;
```

Comment:
Not sure I follow?  `nodesToSize` is used for the actual decision, to store the number of matching bytes for nodes that `canAllocate` returns a YES/THROTTLE decision for.  The `nodeDecisions` are populated regardless of whether they end up in `nodesToSize`, b/c we want the shard store status for every node, even if `canAllocate` returns a NO decision.